clock/testing/fake_clock_test.go

```diff
@@ -194,3 +194,57 @@ func TestFakeStop(t *testing.T) {
 		t.Errorf("expected existing waiter to be cleaned up, but it is still present")
 	}
 }
+
+// This tests the pattern documented in the go docs here: https://golang.org/pkg/time/#Timer.Stop
+// This pattern is required to safely reset a timer, so should be common.
+// This also tests resetting the timer
+func TestFakeStopDrain(t *testing.T) {
+	tc := NewFakeClock(time.Now())
+	timer := tc.NewTimer(time.Second)
+	if !tc.HasWaiters() {
+		t.Errorf("expected a waiter to be present, but it is not")
+	}
+	if !timer.Stop() {
+		t.Errorf("stop should report it stopped the timer")
+	}
+
+	timer = tc.NewTimer(time.Second)
+	tc.Step(2 * time.Second)
+	if timer.Stop() {
+		t.Errorf("stop should report the timer had triggered")
+	} else {
+		<-timer.C()
+	}
+
+	timer.Reset(time.Second)
+	if !tc.HasWaiters() {
+		t.Errorf("expected a waiter to be present, but it is not")
+	}
+	select {
+	case <-timer.C():
+		t.Fatal("got time early on clock; haven't stepped yet")
+	default:
+	}
+	tc.Step(2 * time.Second)
+	<-timer.C()
+}
+
+func TestTimerNegative(t *testing.T) {
+	tc := NewFakeClock(time.Now())
+	timer := tc.NewTimer(-1 * time.Second)
+	if !tc.HasWaiters() {
+		t.Errorf("expected a waiter to be present, but it is not")
+	}
+	go tc.Step(1 * time.Millisecond)
```

Comment:
This is awkward and is required to avoid a deadlock as the fake clock currently works.

I think in an ideal world adding something with 0 or negative would kick off an immediate write to the channel, but that refactor seems a bit more complex, and wasn't actually needed to fix my use-case for this.

Reply:
I updated it to `tc.Step(0)` which seems a bit less awkward, though still not ideal.